File: c++/Graph/Puentes_Itmos.cpp
Algoritmo para hallar los puentes e itsmos en un grafo no dirigido.

vector<int> ady[1010];
int marked[1010];
int previous[1010];
int dfs_low[1010];
int dfs_num[1010];
int itsmos[1010];
int n, e;
int dfsRoot,rootChildren,cont;
vector<pair<int,int>> bridges;

void dfs(int u){
    dfs_low[u] = dfs_num[u] = cont;
    cont++;
    marked[u] = 1; 
    int j, v;

    for(j = 0; j < ady[u].size(); j++){
        v = ady[u][j];
        if( marked[v] == 0 ){
            previous[v] = u;
            //para el caso especial 
            if( u == dfsRoot )  rootChildren++;
            dfs(v);

            //Itsmos
            if( dfs_low[v] >= dfs_num[u] )  itsmos[u]=1;
            
            //Bridges
            if( dfs_low[v] > dfs_num[u] ) bridges.push_back(make_pair(min(u,v),max(u,v)));
            
            dfs_low[u] = min(dfs_low[u], dfs_low[v]);
        }else if( v != previous[u] )  dfs_low[u] = min(dfs_low[u], dfs_num[v]);
    }
}

int main(){
    //Antes de ejecutar el Algoritmo
    cont = dfsRoot = rootChildren = 0;
    bridges.clear();
    dfs( dfsRoot );
    /* Caso especial */
    itmos[dfsRoot] = ( itmos[ dfsRoot ] == 1 && rootChildren > 1 ) ? 1 : 0;
}

File: java/Graphs/Puentes_Itmos.java
Algoritmo para hallar los puentes e itsmos en un grafo no dirigido. Requiere de la clase Edge.

import java.util.ArrayList;
import java.lang.Math;

static int n, e; //vertices, arcos
static int MAX=1010;     
static ArrayList<Integer> ady[]=new ArrayList [MAX];
static boolean marked[]=new boolean [MAX];
static int prev[]=new int [MAX];
static int dfs_low[]=new int [MAX];
static int dfs_num[]=new int [MAX];
static int itsmos[]=new int [MAX];
static ArrayList<Edge> bridges;
static int dfsRoot, rootChildren, cont;

/* Recibe el nodo inicial */
static void dfs(int u){
    dfs_low[u] = dfs_num[u] = cont;
    cont++;
    marked[u] = true;
    int j, v;

    for(j = 0; j < ady[u].size(); j++){
        v = ady[u].get(j);
        if( !marked[v] ){
            prev[v] = u;
            //Caso especial 
            if( u == dfsRoot )  rootChildren++;
            dfs(v);

            //Itmos
            if( dfs_low[v] >= dfs_num[u] )  itsmos[u]=1;
            
            //Puentes
            if( dfs_low[v] > dfs_num[u] )   bridges.add(new Edge( Math.min(u,v),Math.max(u,v)) );
            
            dfs_low[u] = Math.min(dfs_low[u], dfs_low[v]);
        }else if( v != prev[u] )    dfs_low[u] = Math.min(dfs_low[u], dfs_num[v]);
    }
}

public static void main(String args[]){
    //Antes de ejecutar el Algoritmo
    cont = dfsRoot = rootChildren = 0;
    bridges = new ArrayList<Edge>();
    dfs( dfsRoot );
    /* Caso especial */
    itmos[dfsRoot] = ( itmos[ dfsRoot ] == 1 && rootChildren > 1 ) ? 1 : 0;
}

